Doc/c-api/allocation.rst

```diff
@@ -16,7 +16,15 @@ Allocating Objects on the Heap
 
    Initialize a newly allocated object *op* with its type and initial
    reference.  Returns the initialized object.  Other fields of the object are
-   not affected.
+   not initialized.  Specifically, this function does **not** call the object's
+   :meth:`~object.__init__` method (:c:member:`~PyTypeObject.tp_init` slot).
+
+   .. warning::
+
+      This function does not guarantee that the memory will be completely
+      zeroed before it is initialized.  Fields that are not initialized by this
+      function will have indeterminate values, which might include sensitive
+      data from previously destroyed objects (e.g., secret keys).
```

Comment:
I don't think we need such a long note on the consequences of a malloc-like API. It's good to *note* it, as a warning, but warnings are best kept short.
Clearing out secret keys, in particular, should be the job of whoever freed that memory, not the code that reuses it.
```suggestion
   .. warning::

      This function does not guarantee that the memory will be completely
      zeroed before it is initialized.
```